File: scripts/apply_migration.py
#!/usr/bin/env python3
"""Apply SQL migration to Supabase database via REST API."""

import os
import sys
from pathlib import Path

# Add parent directory to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent))

from dotenv import load_dotenv
from supabase import create_client

# Load environment variables
load_dotenv()

SUPABASE_URL = os.getenv("SUPABASE_URL")
SUPABASE_KEY = os.getenv("SUPABASE_SERVICE_KEY")

if not SUPABASE_URL or not SUPABASE_KEY:
    print("Error: SUPABASE_URL and SUPABASE_SERVICE_KEY must be set")
    sys.exit(1)


def apply_migration(sql_file: str) -> bool:
    """Apply a SQL migration file to Supabase."""
    # Read SQL file
    sql_path = Path(sql_file)
    if not sql_path.exists():
        print(f"Error: SQL file not found: {sql_file}")
        return False

    sql_content = sql_path.read_text()
    print(f"Applying migration: {sql_path.name}")
    print(f"SQL length: {len(sql_content)} characters")

    # Note: Supabase client created but not used since supabase-py
    # doesn't support raw SQL execution
    _ = create_client(SUPABASE_URL, SUPABASE_KEY)  # noqa: F841

    # Split SQL into individual statements
    # This is needed because the REST API may not handle multiple statements well
    statements = []
    current = []
    in_function = False

    for line in sql_content.split('\n'):
        stripped = line.strip()

        # Track if we're inside a function definition
        if 'CREATE TRIGGER' in line.upper() or 'CREATE OR REPLACE FUNCTION' in line.upper():
            in_function = True

        current.append(line)

        # Check for statement end (semicolon at end of line, not in function)
        if stripped.endswith(';') and not in_function:
            statement = '\n'.join(current).strip()
            if statement and not statement.startswith('--'):
                statements.append(statement)
            current = []
        elif 'EXECUTE FUNCTION' in line.upper() and stripped.endswith(';'):
            # End of trigger statement
            statement = '\n'.join(current).strip()
            if statement:
                statements.append(statement)
            current = []
            in_function = False
        elif stripped == '$$ LANGUAGE plpgsql;':
            # End of function definition
            statement = '\n'.join(current).strip()
            if statement:
                statements.append(statement)
            current = []
            in_function = False

    # Add any remaining content
    if current:
        statement = '\n'.join(current).strip()
        if statement and not statement.startswith('--'):
            statements.append(statement)

    print(f"Found {len(statements)} SQL statements to execute")

    # Execute each statement via RPC
    # Since we don't have a direct SQL exec, we'll need another approach
    # The supabase-py library doesn't support raw SQL execution
    # We need to use the Supabase SQL Editor API or psycopg2

    print("\nNote: The supabase-py library doesn't support direct SQL execution.")
    print("The migration SQL file has been created at:")
    print(f"  {sql_path.absolute()}")
    print("\nTo apply this migration, you can:")
    print("1. Copy the SQL and paste it in Supabase Dashboard > SQL Editor")
    print("2. Use the Supabase CLI: supabase db push")
    print("3. Connect directly with psql using the connection string from Supabase Dashboard")

    return True


if __name__ == "__main__":
    if len(sys.argv) < 2:
        # Default to the dubbing studio migration
        sql_file = "supabase_migrations/migrations/002_dubbing_studio_schema.sql"
    else:
        sql_file = sys.argv[1]

    success = apply_migration(sql_file)
    sys.exit(0 if success else 1)

File: scripts/apply_migration.mjs
#!/usr/bin/env node
/**
 * Apply SQL migration to Supabase database using the Management API.
 */

import { createClient } from '@supabase/supabase-js';
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Load environment variables from .env file
const envPath = join(__dirname, '..', '.env');
const envContent = readFileSync(envPath, 'utf-8');
const env = {};
for (const line of envContent.split('\n')) {
    const trimmed = line.trim();
    if (trimmed && !trimmed.startsWith('#') && trimmed.includes('=')) {
        const [key, ...valueParts] = trimmed.split('=');
        env[key.trim()] = valueParts.join('=').trim();
    }
}

const SUPABASE_URL = env.SUPABASE_URL;
const SUPABASE_SERVICE_KEY = env.SUPABASE_SERVICE_KEY;

if (!SUPABASE_URL || !SUPABASE_SERVICE_KEY) {
    console.error('Error: SUPABASE_URL and SUPABASE_SERVICE_KEY must be set in .env');
    process.exit(1);
}

// Read the migration SQL
const migrationPath = join(__dirname, '..', 'supabase_migrations', 'migrations', '002_dubbing_studio_schema.sql');
const sql = readFileSync(migrationPath, 'utf-8');

console.log('Migration file:', migrationPath);
console.log('SQL size:', sql.length, 'bytes');

// Create Supabase client
const supabase = createClient(SUPABASE_URL, SUPABASE_SERVICE_KEY);

// The Supabase JS client doesn't support raw SQL execution
// We need to use the postgres.js library directly or the REST API

console.log('\nNote: The Supabase JS client does not support raw SQL execution.');
console.log('The migration SQL file has been created at:');
console.log(' ', migrationPath);
console.log('\nTo apply this migration, you can:');
console.log('1. Copy the SQL and paste it in Supabase Dashboard > SQL Editor');
console.log('2. Use the Supabase CLI with: supabase db execute < migration.sql');
console.log('3. Connect directly with psql using the connection string from Supabase Dashboard');

// Let's try to verify if tables exist
console.log('\nChecking if tables already exist...');

async function checkTables() {
    const tables = ['dub_projects', 'dub_tracks', 'dub_segments'];

    for (const table of tables) {
        try {
            const { data, error } = await supabase.from(table).select('id').limit(1);
            if (error) {
                console.log(`✗ ${table}: ${error.message}`);
            } else {
                console.log(`✓ ${table} exists`);
            }
        } catch (e) {
            console.log(`✗ ${table}: ${e.message}`);
        }
    }
}

checkTables().then(() => {
    console.log('\nDone.');
});
